Unix/protocol/protocol.c

```diff
@@ -440,6 +469,96 @@ static StrandFT _ProtocolSocket_FT = {
     NULL,
     NULL };
 
+/**************** protocolSocketTracker-support **********************************************************/
+static size_t _ProtocolSocketTrackerTestHash(const HashBucket *b)
+{
+    TrackerBucket *bucket = (TrackerBucket*) b;
+    size_t h = 0;
+    Sock key = bucket->key;
+    h = (unsigned)key % 1024;
+    return h;
+}
+
+static int _ProtocolSocketTrackerTestEqual(const HashBucket *b1, const HashBucket *b2)
+{
+    TrackerBucket *bucket1 = (TrackerBucket*) b1;
+    TrackerBucket *bucket2 = (TrackerBucket*) b2;
+    return bucket1->key == bucket2->key;
+}
+
+static void _ProtocolSocketTrackerTestRelease(HashBucket *b)
+{
+    return;
+}
+
+static MI_Result _ProtocolSocketTrackerRemoveElement(Sock s)
+{
+    TrackerBucket b;
+    int r;
+
+    b.key = s;
+
+    Lock_Acquire(&s_trackerLock);
+    r = HashMap_Remove(&s_protocolSocketTracker, (HashBucket*)&b);
+    Lock_Release(&s_trackerLock);
+
+    trace_TrackerHashMapRemove(s);
+
+    return r == 0 ? MI_RESULT_OK : MI_RESULT_FAILED;
+}
+
+static MI_Result _ProtocolSocketTrackerAddElement(Sock s, void *p)
+{
+    TrackerBucket *b;
+    int r;
+
+    b = (TrackerBucket*)PAL_Calloc(1, sizeof(TrackerBucket));
+    if (b == NULL)
+        return MI_RESULT_FAILED;
+
+    b->key = s;
+    b->value = p;
+
+    Lock_Acquire(&s_trackerLock);
+    r = HashMap_Insert(&s_protocolSocketTracker, (HashBucket*)b);
+    Lock_Release(&s_trackerLock);
+
+    if (r != 0)
+    {
+        // already exists
+        trace_TrackerHashMapAlreadyExists(p, s);
+        MI_Result res = _ProtocolSocketTrackerRemoveElement(s);
+        if (res == MI_RESULT_OK)
+        {
+            Lock_Acquire(&s_trackerLock);
+            r = HashMap_Insert(&s_protocolSocketTracker, (HashBucket*)b);
```

Comment:
It could.  And in that case, this AddElement function will return MI_RESULT_FAILED.